### ISSUE TITLE
Brush size keybinding

### ISSUE THREAD

**Issue by RobAnKo** [napari/napari#3994]
## 🚀 Feature

Keybinding to change label paint brush size.
## Motivation

During labeling, I usually completely label each object (cell, cell colony, etc) before moving to another. To do so I label the coarse structure of the object first, using little zoom ("overview mode"). Then, I zoom in to label the finer details of the object (rim areas, branches, string-like features), before I switch to the next object. This involves quite a bit of paint brush size changing. 
Changing the size in the layer controls is inconvenient, as I have to move the cursor out of the image to be labeled, change the size, move back to the image to check if the new size is correct, (optionally move back to the layer controls to re-adjust), and then continue labeling. This disrupts the workflow.

## Pitch

I suggest a simple key binding (e.g. `q` and `e` to have them close to the `1`-`5` mode switches) to adjust the brush size.

## Alternatives

Alternatively, one could change the brush size using the mouse wheel (not used at the moment as far as I know).

## Additional context

This is my first request. I searched the issue list for potential duplicates of this request, but didn't find any, so I hope this request makes sense.
There already is a  menu `Preferences -> Shortcuts -> Group: Labels layer` to adjust keybinding, so I think it should be quite easy to add an option for brush size adjustment.

If anybody has pointers towards where this behavior is controlled, we ([MBPhys](https://github.com/MBPhys/) and me) are happy to implement this ourselves and merge it if you'll allow it.

**Comment by alisterburt** [comment]
Hi @RobAnKo ! Thanks for opening this issue. Really nice idea, I've felt the need for this before too and am surprised there's no issue already for this 🙂 

I feel a slight preference to follow the convention of photoshop and use `[` `]` to change the size rather than `e`/`q` - maybe this isn't as pervasive as I think though, thoughts @jni @kevinyamauchi @tlambert03 ?

In terms of next steps, a PR from yourself and @MBPhys would be a great way to move forward!

The way to go about this would be to implement new `increment_brush_size`/`decrement_brush_size` functions as keybindings following the pattern here
https://github.com/napari/napari/blob/5f96d5d814aad697c367bdadbb1a57750e2114ad/napari/layers/labels/_labels_key_bindings.py#L77-L80

There's some potential complexity here because that we would probably want 'press and hold' of the key to repeatedly increase/decrease the size rather than needing to repeatedly press the key... the only example of press and hold behaviour currently in napari is 'hold space to pan and zoom' which is implemented here

https://github.com/napari/napari/blob/0f5281f641ac9ad468648abdc8b3f30b99ae4b70/napari/layers/labels/_labels_key_bindings.py#L9-L20

See also some info in #3603, I also added some similar behaviour in this PR which may also be useful

https://github.com/napari/napari/pull/3759/files#diff-05d32eb8730c22178a6e17a2db74b2ae8e9a1ff29f855fa2930ef7c05871ce16R36-R59

In the end there may also be some nuances around settings that I'm missing here, hopefully others will chime in on the PR if so and we can figure that out later.

Good luck, don't hesitate to reach out if you have trouble getting this up and running!

**Comment by tlambert03** [comment]
yes, love this.  can't tell you how many times I've pressed `[` or `]` from photoshop muscle memory (I have also found myself using `ctrl`+`option`+drag left/right ... which also works to change brush size in photoshop.  either/both would be awesome :)

**Comment by RobAnKo** [comment]
Thanks for your responses, glad to hear the initial approval :+1: 

Based on what I've seen in the code, to make the customizable shortcut appear in  `Preferences -> Shortcuts -> Group: Labels layer`, we should use the `register_label_action` decorator in napari/layers/labels/_labels_key_bindings.py like so (using a basic function without any hold functionality for now)
```
@register_label_action(
    trans._("Increase the paint brush size by one."),
)
def increase_brush_size(layer: Labels):
    """Increase the brush size"""
    if layer.brush_size < 40: #here we need to add a reference to the limit values of brush size
        layer.brush_size += 1
```
 and add the default keys in napari/napari/utils/shortcuts.py like so 
```
    'napari:increase_brush_size': [']'],
```
Right? 

Also: Can we reference the limit on the the brush sizefrom somewhere instead of hardcoding it as above?

**Comment by tlambert03** [comment]
> Will the new keybinding magically appear in the list of keybindings at `Preferences -> Shortcuts -> Group: Labels layer` or do we need to connect this somehow?

yes, if you follow the pattern for the other default keybindings.  I believe the pattern you'll want is something like
- add 2 functions to `_labels_key_bindings.py` that changes brush size either direction
- decorate them with `@register_label_action`, following the pattern
- add them to `napari.utils.shortcuts.default_shortcuts`

**Comment by kevinyamauchi** [comment]
+1 I would find this very useful!

**Comment by RobAnKo** [comment]
Hey everyone, @MBPhys
we now have a working implementation for key-bound increase/decrease of brush size. :partying_face: 

However, there is a bit of an issue with the hold-to-change-continuously functionality in our solution. To allow for autorepeat, we just added our favorite keys (`Q`, `E`, but could be any pair) to the list of special keys in line 482 of key_bindings.py

https://github.com/napari/napari/blob/6a3e11aa717e7928a0a5a3c7693577729a466ef1/napari/utils/key_bindings.py#L471-L482

So that's a little hacky. It works as long as users stick to the default. When changing the brush size keybinding in `Preferences -> Shortcuts -> Group: Labels layer`, increase/decrease still works with the new key, but not continuously if the new key is not in `['Up', 'Down', 'Left', 'Right', 'Q', 'E']`.

This made us wonder: Would it make sense to add a boolean flag to ALL keybindings, which 
- defines if this function allows autorepeat
- is used to create a list of all autorepeat-able keys. This list could be referenced as exceptions in key_bindings instead of hard-coded keys.
- can be changed by the user in `Preferences -> Shortcuts` (an additional column 'autorepeat' in the menu with a simple checkbox)

As far as we understand, all key-bound callbacks are executed through the on_key_press method, which restricts them all to be executed only once, except for the navigation keys. Since there is probably a good reason to do that, we thought that adding additional flexibility through the autorepeat flag could be a useful pattern for this specific issue, but also for potential future extensions.

There might be unintended side effects / instabilities arising from this approach that we are currently not aware of, so please feel free to roast this if you have reasons to believe this could lead to problems.

**Comment by alisterburt** [comment]
Amazing @RobAnKo ! Can't wait to get this in 🥳 🚀 

I think adding a boolean flag to the `bind_key` method sounds like a great solution - I'm sure some complexity I don't know about will come up but there's only one way to find out! Feel free to ping me here if you run into trouble but it sounds like you're well on your way :)

FYI - we run community meetings where we discuss/demo recent developments in napari, would you and @MBPhys be interested in joining one? The timetable is available [here](https://napari.org/community/meeting_schedule.html), we would love to hear about what you're up to!

**Comment by RobAnKo** [comment]
Thanks for the invite! We already connected with some folks on zulip to discuss plugins that we (mainly @MBPhys) developed. I'll definitely join on Wednesday, Marc probably too, depending on his schedule.

**Comment by alisterburt** [comment]
Look forward to seeing you there!

**Comment by psobolewskiPhD** [comment]
Bump, this feature would be super useful.
@RobAnKo are you still thinking of a PR for this?
Seems like at the moment the path of least resistance is to hard-code `[` and `]` (the Photoshop keybinds). 
Maybe perfect shouldn't get in the way of good?
CC: @alisterburt @tlambert03 @jni

**Comment by alisterburt** [comment]
Big +1 for something rather than nothing, even if it's a little clunky!

**Comment by psobolewskiPhD** [comment]
This scroll wheel binding from @ianhi is pretty sweet too:
https://github.com/napari/napari/issues/4837#issuecomment-1189342980
https://github.com/Hekstra-Lab/napari-labeller/blob/74913dce72c773df2ec94e1cb3798dd40fedf219/src/napari_labeller/_keybinds.py#L59-L66

Though personally I might prefer the scroll wheel just did zoom independent of PAN on the Labels layer...

**Comment by psobolewskiPhD** [comment]
@alisterburt I quickly implemented the hard-coding method from above in a branch, using [ and ].
https://github.com/psobolewskiPhD/napari/tree/feature/keybind_label_brush_size
It's pretty life-altering....
😍

**Comment by RobAnKo** [comment]
Hi folks,
thanks for reminding me of this. 
I implemented the hard-coding method @psobolewskiPhD referred to locally and since it works for me, I focussed on other non-napari things since then :see_no_evil: 
I'll dig back into this as soon as I find some spare time and will implement the ideas mentioned above https://github.com/napari/napari/issues/3994#issuecomment-1025713296.

@alisterburt Since I am not a regular GitHub user and unfamiliar with PRs, I'll probably reach out to ask about the process of creating the pull request once I have something that works.

**Comment by alisterburt** [comment]
@RobAnKo it would be a pleasure to help you get familiar with the GitHub flow, feel free to ping me here/on zulip any time!

**Comment by psobolewskiPhD** [comment]
Hi @RobAnKo 
I took the liberty of making a WIP PR with the hard coding.
That way it's easier for people to try it out and we can find a good solution.

**Comment by RobAnKo** [comment]
Thanks a lot to both of you. 
@psobolewskiPhD Now I'm not sure on how to proceed when I make changes. Can I push changes to your PR directly? 
Also: should we move communication to the Conversation on your PR?

**Comment by alisterburt** [comment]
Let's move over to peters PR - for future reference, you can PR his PR but not push to it directly

**Comment by psobolewskiPhD** [comment]
You can just open your own and I will gladly close mine--I just wanted something up so people could check-out and try it; it's life changing!

Alternatively it should be possible to push to that branch, but honestly I'm not quite sure how. Paging @alisterburt

**Comment by RobAnKo** [comment]
It is life changing indeed! Let's keep your PR open, as it already describes the issue perfectly and presents a minimum working solution for people to experience the pure bliss of auto-repeating brush size key bindings :heart:
I'll quote my above-mentioned idea for a general solution of press-and-hold keybindings in the PR so that the information is found in one place.
